Enhance for including headers on libraries / Tests / Packages with samples (#2088)
Azure/azure-sdk: docs/cpp/implementation.md

{% include requirement/MUSTNOT id="cpp-style-change-headers" %} substantially change the names exposed by the header in response to macros or other controls. For example, `NOMINMAX` or `WIN32_LEAN_AND_MEAN` from `<Windows.h>`.

> {% include requirement/SHOULD id="cpp-design-physical-include-quotes" %} include files using quotes (") for files within the package's library, and angle brackets (<>) for external dependencies. Consider unit tests, samples and any other code out of the package's library to be a reference for how the package's library is consumed, hence, include headers using angle brackets (<>) on this cases.

Reviewer: ```suggestion
{% include requirement/SHOULD id="cpp-design-physical-include-quotes" %} include files using quotes (") for files defined within the package, and angle brackets (<>) for including files defined externally. Consider unit tests, samples and any other code out of the package's library to be a reference for how the package's library is consumed, hence, include headers using angle brackets (<>) for such cases.
```

{% highlight cpp %}

> // Including headers from within a package library like Azure Core

Reviewer: nit: Saying both `package` and `library` seems redundant.
```suggestion
// Including headers from within a package like Azure Core
```

#include "azure/core/context.hpp"
#include "azure/core/http/http.hpp"

#include <iostream>  // external
#include <vector>

> // Including headers from the same package but outside the library (like tests or samples)

Reviewer: nit: Since we don't include tests (or samples) as part of the package, it is a bit confusing to say `"from the same package but outside the library"`

Maybe:

```suggestion
// Including headers from a library within tests and samples for that specific library at the same package-level directory.
```

#include <azure/core/context.hpp>
#include <azure/core/http/http.hpp>

#include <iostream>  // external
#include <vector>

// Including headers from another library (like Azure Storage including Azure Core)
#include <azure/core/context.hpp>
#include <azure/core/http/http.hpp>

> #include "azure/storage/blobs.hpp" // same package's library

Reviewer: ```suggestion
#include "azure/storage/blobs.hpp" // including header files defined in the same library (in this case, Azure Storage)
```
